### ISSUE TITLE
Build Issues with Go Template for Raspberry Pi 3Bs

### ISSUE THREAD

**Issue by markserrano915** [openfaas/templates#135]
I got two problems whenever I try to build a Go function for Raspberry Pi 3Bs.

To build the project in MacOS, I did the following:
```
faas-cli new pet-api --lang go-armhf
faas-cli build -f pet-api.yml
```

However when I run the build, I get the following error:

> The command '/bin/sh -c test -z ./main.go' returned a non-zero code: 1
> 2019/03/13 14:36:25 ERROR - Could not execute command

The offending section is:

https://github.com/openfaas/templates/blob/45db7b49fb4c3a7d024ba96d0c12aef6dd5ecc86/template/go-armhf/Dockerfile#L13

I don't know what `-z` does in `test -z`. But I assumed it's testing for the presence of file. Google showed that `test -e` will do the trick. So I replaced it with:

```
RUN test -e "$(gofmt -l $(find . -type f -name '*.go' -not -path "./vendor/*" -not -path "./function/vendor/*"))" || { echo "Run \"gofmt -s -w\" on your Golang code"; exit 1; }
```
That worked.

Now I deployed the project via
```
faas-cli deploy -f pet-api.yml
```
Then I got hit with this error upon calling the function:

> fork/exec
> ./handler: exec format error

So it turns out the offending line is this:

https://github.com/openfaas/templates/blob/45db7b49fb4c3a7d024ba96d0c12aef6dd5ecc86/template/go-armhf/Dockerfile#L15

I have to compile it for ARM as follows:
```RUN CGO_ENABLED=0 GOOS=linux GOARCH=arm GOARM=7 \
    go build --ldflags "-s -w" -a -installsuffix cgo -o handler . && \
    go test ./... -cover
```

When I run the function, I was able to get the info I need.

Also upgrading to the latest arm version works as well:

```
FROM arm32v7/golang:1.12.0-alpine3.9 
```

I am thinking of creating a PULL request.

**Comment by alexellis** [comment]
Hi thanks for your interest in the project.

Please don't raise a PR.

go-armhf will not work on MacOS. That is only for on-device compilation. For MacOS use simply the "go" template. 

If "go-armhf" has developed an issue on the device then that is something that we need to look into. 

I'm going to close this issue because I don't think there is a problem. I will be open it again if needed but you can keep commenting here. 

Alex

**Comment by markserrano915** [comment]
I'll try this later. 

So my target platform are Raspberry Pi's. I will make a Go (non-ARM) template in Mac. Build it and deploy to OpenFaas on my Raspberry Pi cluster. I kinda expect that won't work since I thought Raspberry Pi's would need a GO build for ARM?

Well I think I tried that already because the original template (https://github.com/openfaas/templates/blob/master/template/go-armhf/Dockerfile) is building the project without these flags: `GOARCH=arm GOARM=7`. (My previous post didn't show that I added these flags)
